rust/xaynet-server/src/settings/mod.rs

```diff
@@ -0,0 +1,623 @@
+//! Loading and validation of settings.
+//!
+//! Values defined in the configuration file can be overridden by environment variables. Examples of
+//! configuration files can be found in the `configs/` directory located in the repository root.
+
+use config::{Config, ConfigError, Environment};
+use redis::{ConnectionInfo, IntoConnectionInfo};
+use serde::{
+    de::{self, Deserializer, Visitor},
+    Deserialize,
+};
+use std::{fmt, path::PathBuf};
+use thiserror::Error;
+use tracing_subscriber::filter::EnvFilter;
+use validator::{Validate, ValidationError, ValidationErrors};
+
+use xaynet_core::mask::{BoundType, DataType, GroupType, MaskConfig, ModelType};
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(not(feature = "model-persistence"))]
+#[derive(Debug, Validate, Deserialize, Default)]
+pub struct S3Settings {}
+#[cfg(feature = "model-persistence")]
+pub mod s3;
+#[cfg(feature = "model-persistence")]
+pub use self::{s3::S3BucketsSettings, s3::S3Settings};
+
+#[derive(Error, Debug)]
+/// An error related to loading and validation of settings.
+pub enum SettingsError {
+    #[error("configuration loading failed: {0}")]
+    Loading(#[from] ConfigError),
+    #[error("validation failed: {0}")]
+    Validation(#[from] ValidationErrors),
+}
+
+#[derive(Debug, Validate, Deserialize)]
+/// The combined settings.
+///
+/// Each section in the configuration file corresponds to the identically named settings field.
+pub struct Settings {
+    pub api: ApiSettings,
+    #[validate]
+    pub pet: PetSettings,
+    pub mask: MaskSettings,
+    pub log: LoggingSettings,
+    pub model: ModelSettings,
+    #[validate]
+    pub metrics: MetricsSettings,
+    pub redis: RedisSettings,
+    #[validate]
+    #[cfg_attr(not(feature = "model-persistence"), serde(default))]
+    pub s3: S3Settings,
```

Comment:
as suggested above, we could gate this entirely behind the corresponding feature flag as follows:
```suggestion
    #[cfg(feature = "model-persistence")]
    #[validate]
    pub s3: S3Settings,
```